Rework organization creation flow.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First three commits are prep commits for renaming and extracting some common code.
- Next commit is to add `PreregistrationRealm` class.
- Next 6 commits are to update the code to use `PreregistrationRealm` objects for realm creation. These all commits will be squashed, just kept them separate for now so that it is easier to review.
- Last commit is for UI changes in the user registration form with realm details pre-filled.

Fixes part of #24307.
 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Screenshot from 2023-03-15 23-28-14](https://user-images.githubusercontent.com/35494118/225401151-cf03ec5e-93db-4e1f-83eb-e0f7e1c449f1.png)
![Screenshot from 2023-03-14 23-39-21](https://user-images.githubusercontent.com/35494118/225102865-b2900a5f-6544-463f-a6e4-14048071ce5b.png)
![realm-creation-new](https://user-images.githubusercontent.com/35494118/225102890-5f278e32-7813-4260-b74d-ab8aac9be830.gif)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
Posting some of the initial questions in case someone sees this today -
- How we should handle the existing `PreregistrationRealm` objects with `string_id` (subdomain) for which a realm has been created? I guess we would want to revoke them but do we want to revoke even if it was created by some other user or we just want to keep the link active and then prompt user during registration that subdomain is already used.
- If user modifies the realm name, type or subdomain in the second form, i.e. the user registration form, then do we want to update the `PreregistrationRealm` objects with the new fields or we just want to keep the old fields. I have added the `created_realm` and `created_user` field to the `PreregistrationRealm` fields for now to track the object which was used for creating the realm.
- How should we handle the already existing `PreregistrationUser` objects created for realm creation. They may not work now witht the current implementation. Do we just want to revoke them?

sahil839 commented:
Will probably post a couple of more questions regarding code tommorrow.

alya commented:
The screenshots look good!

I'd change the "Email" field label on the first form to "Your email" for clarity, since everything else it's asking is about the org.  Just "Email" seems fine to me for the second form.

sahil839 commented on zerver/views/auth.py:
```diff
@@ -120,18 +120,13 @@ def get_safe_redirect_to(url: str, redirect_host: str) -> str:
 def create_preregistration_user(
     email: str,
     realm: Optional[Realm],
```
Here we can change the type to `Realm`, as we always pass realm to this function. Have kept this as `Optional[Realm]` for now as realm field passed [here] has type defined as `Optional[Realm]` [here](https://github.com/zulip/zulip/blob/d342f704406ba68a2e41ea7cf32744d2a43d88b4/zerver/views/auth.py#L190).
I am sure we were passing realm object and not None from the above linked code as otherwise the below assert statement (which is now removed) would raise an error.

Can we instead raise error in the try-except block [here](https://github.com/zulip/zulip/blob/d342f704406ba68a2e41ea7cf32744d2a43d88b4/zerver/views/auth.py#L190)? Because if we set realm as `None` in the except block, then `create_preregistration_user` would have eventually raised an `AssertionError`. I am not much familiar with this code but just pointing it out.

sahil839 commented:
I don't have questions in the code currently, but the code in `registration_helper` is fairly complicated and thus needed to add a few `assert isinstance` statements to pass lint checks and would be good to probably refactor it a bit in future.
Also, have used a couple of functions for handling both `PreregistrationUser` and `PreregistrationRealm` objects currently, which makes code somehwat confusing but since we can change the realm details in second form as well while creating user, it becomes complicated to completely separate the code.

Just to note that I posted a few questions [here](https://github.com/zulip/zulip/pull/24689#issuecomment-1468636478).

Also, we can now probably remove the `realm_creation` field from `PreregistrationUser` class. Can do it separately in a new PR or once this PR is reviewed once.

sahil839 commented:
> I'd change the "Email" field label on the first form to "Your email" for clarity, since everything else it's asking is about the org. Just "Email" seems fine to me for the second form.

Updated this.

I am not sure why the production suite is failing though. I already tried rerunning it once.

@alya would be good if you can do a manual testing on this once.

timabbott commented on zerver/forms.py:
```diff
@@ -262,6 +262,19 @@ def email_is_not_disposable(email: str) -> None:
+            subdomain = Realm.SUBDOMAIN_FOR_ROOT_DOMAIN
+
+        check_subdomain_available(subdomain)
+        return subdomain
```
Does it make sense to change the registration form to subclass this or something to avoid duplication of the checks for fields that appear in both?

sahil839 replied:
I have tried using a new class `RealmDetailsForm` which is used as subclass by both `RealmCreationForm` and `RegistrationForm`.

I have not directly used `RealmCreationForm` as subclass for `RegistrationForm` since `email` is not a field in `RegistrationForm` and we would need to set `email = None` in `RegistrationForm` in that case.

timabbott commented on zerver/views/registration.py:
```diff
@@ -701,7 +701,16 @@ def create_realm(request: HttpRequest, creation_key: Optional[str] = None) -> Ht
+            "root_domain_available": is_root_domain_available(),
+            "sorted_realm_types": sorted(
+                Realm.ORG_TYPES.values(), key=lambda d: d["display_order"]
+            ),
```
Similar question here -- should we have a function `get_realm_create_form_context` to do the most of this that should be shared?

timabbott commented on zerver/tests/test_management_commands.py:
```diff
@@ -317,7 +325,15 @@ def test_generate_link_confirm_email(self) -> None:
+                "realm_type": Realm.ORG_TYPES["business"]["id"],
+                "realm_subdomain": "zuliptest",
+            },
+        )
```
Do these test changes belong in the previous commit, since it's the one that extends the form?

sahil839 replied:
Yes, I thought the same. And yes we would end up squashing 6 commits in the middle, as I mentioned in the PR description.

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -3849,7 +3941,15 @@ def test_create_two_realms(self) -> None:
+                "realm_type": Realm.ORG_TYPES["business"]["id"],
+                "realm_subdomain": second_string_id,
+            },
+        )
```
Should we be adding a new `test_classes.py` helper function to reduce duplication here? We'd perhaps need an optional kwarg for the `realm_type`

`self.submit_realm_creation_form(email, subdomain=second_string_id, name=second_realm_name)` might feel nicer, ? Idunno curious for thoughts from @mateuszmandera on this.

sahil839 replied:
Added a commit for extracting a helper. Have kept this commit at last for now, since we would squash some of the commits later.

timabbott commented on zerver/actions/create_user.py:
```diff
@@ -228,24 +228,20 @@ def process_new_human_user(
-        if prereg_user is not None and prereg_user.referred_by is not None:
-            notify_invites_changed(user_profile.realm)
+    if prereg_user is not None and prereg_user.referred_by is not None:
+        notify_invites_changed(user_profile.realm)
```
For these blocks, is the point that these filters will be the empty set and `prereg_user.referred_by` will always be `None` in realm creation?

mateuszmandera replied:
Hmm yeah I don't fully understand the exact reasoning and correctness justification of this change - perhaps more elaboration should be added in the commit message?

sahil839 replied:
Yes, the `prereg_user` will be None during realm creation and for the `else` block there will not be any `PreregistrationUser` objects in the realm.
Added the explaination in commit message as well.

timabbott commented on web/src/portico/signup.js:
```diff
@@ -284,4 +284,10 @@ $(() => {
+    $("#registration").on("click", ".edit-realm-details", (e) => {
+        $("#registration .not-editable-realm-details").addClass("hide");
+        $("#registration .realm-creation-editable-inputs").removeClass("hide");
+        $(e.target).hide();
```
Does this do something good with keyboard focus? I could imagine a `preventDefault` being useful as part of making that work reliably.

sahil839 replied:
I am not sure how adding `preventDefault` would help here.

Currently the focus is on "Full name" input, when the form is first opened. When user clicks on pencil icon, none of the inputs is focussed. We might need to focus it on the "Organization name" input in that case.

Also, the pencil icon is not focusable right now. We should make it focusable and might also want a handler to opend the editable form on pressing `Enter` when the icon is in focus.

timabbott commented on templates/zerver/realm_creation_form.html:
```diff
@@ -1,66 +1,70 @@
+        <div class="help-text">
+            {{ _('Shorter is better than longer.') }}
+        </div>
+        {% endif %}
```
What's the thinking behind hiding this help text in this context?

sahil839 replied:
I removed this since it is mentioned in the issue. But on reading the issue again, I am confused whether Alya meant to hide it in the non-editable state and show it when user wants to edit it. @alya can you confirm?

timabbott commented:
This looks great -- posted everything I noticed in a read through everything. I didn't look at the post-review code for how nice it feels, just the diffs, so there may be things I'm missing. @mateuszmandera can you do a thorough review of this? 

> How we should handle the existing PreregistrationRealm objects with string_id (subdomain) for which a realm has been created? I guess we would want to revoke them but do we want to revoke even if it was created by some other user or we just want to keep the link active and then prompt user during registration that subdomain is already used.

Good question. I think what I'd expect as a user would be that trying to create the realm in the second form would give an error that the subdomain was taken (it must have been very recently!) and then require them change that error field in the second form. (So we might need an option to have the second form open in the editable state for this error handling case).

> If user modifies the realm name, type or subdomain in the second form, i.e. the user registration form, then do we want to update the PreregistrationRealm objects with the new fields or we just want to keep the old fields. I have added the created_realm and created_user field to the PreregistrationRealm fields for now to track the object which was used for creating the realm.

Probably a question to discuss in Zulip, but I think either option is defensible.

> How should we handle the already existing PreregistrationUser objects created for realm creation. They may not work now with the current implementation. Do we just want to revoke them?

I'm not sure. Revoking them all is certainly the simplest option. There's a fairly short window around this upgrade during that would have to overlap with when someone was on their way to click the confirmation link for this to come up, so it's possible just doing the deployment for this transition at a weird hour may be good enough. But if we can easily do a temporary commit that we can deploy to Zulip Cloud and drop a week later that just has the second form start in the "Realm creation section open" state for this class of link, or has these old links redirect you to the updated form with the email field prefilled (which at least looks like you made progress), that might be easier to roll out.

mateuszmandera commented on zerver/views/development/registration.py:
```diff
@@ -92,8 +95,11 @@ def register_demo_development_realm(request: HttpRequest) -> HttpResponse:
-    prereg = create_preregistration_user(email, None, realm_creation=True, password_required=False)
-    activation_url = create_confirmation_link(prereg, Confirmation.REALM_CREATION)
+    realm_subdomain = realm_name
+    prereg = create_preregistration_realm(email, realm_name, realm_subdomain, realm_type)
```
These should probably be named `prereg_realm` rather than carrying over the old mistake of naming `prereg` instead of the usual `prereg_user` convention?

sahil839 replied:
Updated it. We still need to rename `prereg` to `prereg_user` for other existing cases. Will do it as a follow-up.

mateuszmandera reviewed (commented):
Reviewed most of the changes and made comments, didn't read all the `test_signup` tests and the html form changes fully yet though

alya commented:
Looks good to me in manual testing!

alya commented:
One thing I noticed (not a regression in this PR) is that the thin down-angle in the org type dropdown doesn't look great to me. Is there a better style we use elsewhere that we could apply here?

sahil839 commented on zerver/views/registration.py:
```diff
@@ -316,24 +345,29 @@ def registration_helper(
                 name_validated = True
                 form = RegistrationForm(
-                    {"full_name": prereg_user.full_name}, realm_creation=realm_creation
+                    {"full_name": prereg_user.full_name},
```
I think this does not use `initial` intentionally, so that the form is validated directly, like we do for LDAP case.

mateuszmandera replied:
Yeah i think that's right, this wasn't using `initial`

sahil839 commented on analytics/tests/test_support_views.py:
```diff
@@ -302,7 +302,15 @@ def get_check_query_result(
             check_zulip_realm_query_result(result)
 
```
Fixed it.

sahil839 commented on zproject/backends.py:
```diff
@@ -992,19 +995,21 @@ def get_or_build_user(self, username: str, ldap_user: _LDAPUser) -> Tuple[UserPr
             opts["default_stream_groups"] = []
 
+        if self._prereg_realm:
+            opts["prereg_realm"] = self._prereg_realm
```
The [registration: Modify PreregistrationRealm objects after registration.](https://github.com/zulip/zulip/pull/24689/commits/c0f334203b973e7a6e467cd72faba66e89c087a9) adds `prereg_realm` to `do_create_realm` but [backends: Update ZulipLDAPAuthBackend to use PreregistrationRealm.](https://github.com/zulip/zulip/pull/24689/commits/113142cbd082ac301d80173ef7e0f3d47f67e2b9) does not use it and it actually uses `prereg_realm` in `authenticate` which is added in this commit itself.

Though, there was some other mistake as [backends: Update ZulipLDAPAuthBackend to use PreregistrationRealm](https://github.com/zulip/zulip/pull/24689/commits/113142cbd082ac301d80173ef7e0f3d47f67e2b9) did not add `prereg_realm` kwarg to `do_create_user`, so change the order anyways.

sahil839 commented on zerver/views/registration.py:
```diff
@@ -296,13 +314,24 @@ def registration_helper(
             # complete (that is, no additional fields need to be
             # filled out by the user) we want the form to validate,
             # so they can be directly registered without having to
             # go through this interstitial.
```
Updated it.

sahil839 commented:
Updated the PR as per feedback in inline comments and also replied to some of the comments. 
Still need to work on (https://github.com/zulip/zulip/pull/24689#discussion_r1139173118) and https://github.com/zulip/zulip/pull/24689#discussion_r1137524561.

> One thing I noticed (not a regression in this PR) is that the thin down-angle in the org type dropdown doesn't look great to me. Is there a better style we use elsewhere that we could apply here?

Will look into this but would probably be better to do in a separate PR.

sahil839 commented:
Regarding questions discussed in https://github.com/zulip/zulip/pull/24689#issuecomment-1472514239:
- Updated the PR to not revoke the `PreregistrationRealm` object with same subdomains, but don't think it is trivial to open the form with realm fields in editable state and error shown becasue we pass the realm details to the form as `initial` data and those values are not considered while validation. And if check the validity of whole form everytime while opening the second form, then it will show errors everytime due to full name field being empty which will be expected. Currently, what I have done is that the form opens with realm fields as non-editable, but when user tries to submit the form with a `string_id` which is already used, then the realm fields become editable with the error shown.
- I will start the discussion about editing `PreregistrationRealm` objects on CZO.
- Will try to do a commit which can be deployed briefly to handle old `PreregistrationUser` objects.

timabbott commented:
(I'm assuming @mateuszmandera will do the next review on this.)

mateuszmandera commented on zerver/views/registration.py:
```diff
@@ -528,6 +563,15 @@ def registration_helper(
         realm_user_default = RealmUserDefault.objects.get(realm=realm)
         default_email_address_visibility = realm_user_default.email_address_visibility
 
+    selected_realm_type_name = None
+    if realm_creation:
+        assert prereg_realm is not None
+        selected_realm_type_name = [
+            org_type["name"]
+            for org_type in Realm.ORG_TYPES.values()
+            if org_type["id"] == prereg_realm.org_type
+        ][0]
```
Should we move this to a helper? This function is already huge and nearly impossible to mentally parse, so seems nice to encapsulate additional fiddly logic elsewhere :neutral_face:

timabbott replied:
Yeah that's a great idea.

sahil839 replied:
Added a helper. We can also directly add the below field to context object instead of adding a new function -
```python
"selected_realm_type_name": None if prereg_realm is None else get_selected_realm_type_name(prereg_realm),
```
Can do this change, if we find it easily readable.

mateuszmandera reviewed (commented):
The backend code looks good to me now! Haven't review the js changes since I don't feel confident reviewing our javascript

timabbott commented:
> The backend code looks good to me now! Haven't review the js changes since I don't feel confident reviewing our javascript

OK; I remember being pretty happy with the JS.  @sahil839 send me a PM when you have this ready for me -- I kinda want to get it integrated while it's fresh in my mind.

sahil839 commented:
Updated the PR. Have also replied some of the comments above already. Other Changes after the last update -
- Added code to make the pencil icon focusable and now the edit form is opened on pressing Enter when pencil icon is in focus. Also, the realm name field is now focussed after the realm details form is opened.

timabbott commented:
I added this docstring for the new class, and also moved it before `PreregistrationUser` just to keep those closer together:
``` diff
+    """Data on a partially created realm entered by a user who has
+    completed the "new organization" form. Used to transfer the user's
+    selections from the pre-confirmation "new organization" form to
+    the post-confirmation user registration form.
+
+    Note that the values stored here may not match those of the
+    created realm (in the event the user creates a realm at all),
+    because we allow the user to edit these values in the registration
+    form (and in fact the user will be required to do so if the
+    `string_id` is claimed by another realm before registraiton is
+    completed).
+    """
```

timabbott commented:
This is great, merged, thanks @sahil839! @alya where do you think we should post about this change in chat.zulip.org asking for contributors to do some extra manual testing on it in the development environment? The new organization creation experience is one of those things that's pretty impactful if it's got bugs, so I'd love to have a few folks hammer on it over the course of this week before it goes to production.

> And yes, we can add a temporary commit for the old PreregistrationUser objects created for realm creation. We would just basically need to alter/remove some assertion statements and some liens of code to show the realm details form always in that

Let's do that in a separate PR, since we won't want to merge it to `main`, and it'll be nice to let GitHub merge this whole branch.

timabbott commented:
The test failure looks like an unrelated test flake; I've triggered a rerun in CI and also am running it locally.
